csrc/val_graph.cpp

```diff
@@ -0,0 +1,463 @@
+  NVF_ERROR(val_group, "Nullptr not allowed");
+  if (auto it = unique_uses_.find(val_group); it != unique_uses_.end()) {
+    return &(it->second);
+  } else {
+    return nullptr;
+  }
+}
+
+void ValGraph::mapVals(Val* val0, Val* val1) {
+  if (val0 == val1) {
+    return;
+  }
+
+  if (disjointValSets().strictAreMapped(val0, val1)) {
+    return;
+  }
+  // Definitions and uses are based on the groups of id0 and id1, don't merge
+  // them into a single group until we grab all definitions and uses for later
+  // processing.
+  ValGroup orig_val_group0 = toGroup(val0);
+  ValGroup orig_val_group1 = toGroup(val1);
+  const ExprGroups* orig_defs0 = getDefinitions(orig_val_group0);
+  NVF_ERROR(orig_defs0);
+  const ExprGroups* orig_defs1 = getDefinitions(orig_val_group1);
+  NVF_ERROR(orig_defs1);
+  const ExprGroups* orig_uses0 = getUses(orig_val_group0);
+  NVF_ERROR(orig_uses0);
+  const ExprGroups* orig_uses1 = getUses(orig_val_group1);
+  NVF_ERROR(orig_uses1);
+
+  // Map the iter domains together before we traverse across definitions and
+  // uses. Traversing definitions and uses could use the new property of id0 and
+  // id1 being mapped.
+  disjointValSets().mapEntries(val0, val1);
+  auto new_val_group = toGroup(val0);
+
+  unique_definitions_[new_val_group] = orig_defs0->computeUnion(*orig_defs1);
+  unique_uses_[new_val_group] = orig_uses0->computeUnion(*orig_uses1);
+
+  // Propagate on uses
+  if (!orig_uses0->empty() && !orig_uses1->empty()) {
+    for (const ExprGroup& use_group_1 : *orig_uses1) {
+      for (const ExprGroup& use_group_0 : *orig_uses0) {
+        if (use_group_0 == use_group_1) {
+          continue;
+        }
+        Expr* use0 = use_group_0->front();
+        Expr* use1 = use_group_1->front();
+        maybeMapThroughExprs(use0, use1, true);
+      }
+    }
+  }
+
+  // Propagate on definitions
+  if (!orig_defs0->empty() && !orig_defs1->empty()) {
+    for (const ExprGroup& def_group_1 : *orig_defs1) {
+      for (const ExprGroup& def_group_0 : *orig_defs0) {
+        if (def_group_0 == def_group_1) {
+          continue;
+        }
+        auto def0 = def_group_0->front();
+        auto def1 = def_group_1->front();
+        maybeMapThroughExprs(def0, def1, false);
+      }
+    }
+  }
+
+  unique_definitions_.erase(orig_val_group0);
+  unique_definitions_.erase(orig_val_group1);
+  unique_uses_.erase(orig_val_group0);
+  unique_uses_.erase(orig_val_group1);
+}
+
+void ValGraph::maybeMapThroughExprs(Expr* expr0, Expr* expr1, bool forward) {
+  if (!exprsMap(expr0, expr1, forward)) {
+    return;
+  }
+
+  // Expr inputs are mapped. If propagate_exprs_ is true, map the
+  // exprs and outputs
+  if (propagate_through_exprs_) {
+    mapExprs(expr0, expr1);
+    mapThroughExpr(expr0, expr1, forward);
+  } else if (
+      inputGroups(toGroup(expr0)) == inputGroups(toGroup(expr1)) &&
+      outputGroups(toGroup(expr0)) == outputGroups(toGroup(expr1))) {
```

Comment:
I don't quite follow. When `propagate_through_exprs_` is off, we check expr0 and expr1 have the same input/output groups. However, when `propagate_through_exprs_` is on, we don't check that any more? I thought `propagate_through_exprs_` only changes whether to propagate rather than how to check individual expression pairs. Am I missing something?

Reply:
By default, expressions are mapped only when everything is matched, i.e., inputs, outputs and attributes are all mapped or equal. 

When the propagation is allowed, as long as the inputs are mapped and the attributes are equal, we propagate the mappings to the outputs and the expressions. 

In either case, it should be always true that when two expressions are mapped, their inputs and outputs are also mapped, respectively, and vice versa.

Reply:
Got it. So should the code be: 

```
  // By default, expressions are mapped only when everything is matched, i.e., inputs, outputs and attributes are all mapped or equal.
  // When the propagation is allowed, as long as the inputs are mapped and the attributes are equal, we propagate the mappings to the outputs and the expressions.
  // In either case, it should be always true that when two expressions are mapped, their inputs and outputs are also mapped, respectively, and vice versa.
  if (inputGroups(toGroup(expr0)) != inputGroups(toGroup(expr1)) {
    return;
  }
  if (propagate_through_exprs_) {
    mapExprs(expr0, expr1);
    mapThroughExpr(expr0, expr1, forward);
  } else {
    if (outputGroups are equal) {
      mapExprs(expr0, expr1);
    }
  }
```

I don't know how different this is in practice. Maybe other code inside `mapExprs` and `mapThroughExpr` does the same check. But I **think** this is more intuitive from readers' perspective.

Reply:
Actually, I believe 

```
inputGroups(toGroup(expr0)) == inputGroups(toGroup(expr1))
```

is redundant when `forward` is true. `exprsMap` must have already checked the inputs.

@csarofeen Do you agree?

Reply:
Yes, and `outputGroups(toGroup(expr0)) == outputGroups(toGroup(expr1))` is redundant when forward is false. I think the logic as written makes sense.

Reply:
Actually, should we move this to `mapVals`, because this operation is not "map through expr", but just do some book keep of exprs whose inputs and outputs is already mapped.

Reply:
Can we please wrap up the nit's. Writing code is rewriting code and we can come back and do further cleanup once this code is functional for the purpose we're building it for.

Reply:
I see -- I missed the `forward` flag. Either the inputGroups check or the outputGroups check is redundant depending on the flag. So it's at least functionally correct to check both.

Reply:
Updated with the comments